src/kernels/MIOpenInstanceNorm.cpp

```diff
@@ -0,0 +1,317 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+#ifndef IN_OUT_TYPE
+#define IN_OUT_TYPE float
+#endif
+
+#ifndef LOCAL_SIZE
+#define LOCAL_SIZE 256
+#endif
+
+template <typename TIO>
+__device__ void instanceNormFwdTrain(const TIO* x,
+                                     TIO* y,
+                                     const TIO* scale,
+                                     const TIO* bias,
+                                     const TIO* running_mean_in,
+                                     const TIO* running_var_in,
+                                     TIO* running_mean_out,
+                                     TIO* running_var_out,
+                                     TIO* mean_var,
+                                     float eps,
+                                     float momentum,
+                                     uint64_t outer_size,
+                                     uint64_t inner_size,
+                                     tensor_view_t<5> x_tv,
+                                     tensor_view_t<5> y_tv,
+                                     tensor_view_t<1> scale_tv,
+                                     tensor_view_t<1> bias_tv,
+                                     tensor_view_t<1> running_mean_in_tv,
+                                     tensor_view_t<1> running_var_in_tv,
+                                     tensor_view_t<1> running_mean_out_tv,
+                                     tensor_view_t<1> running_var_out_tv,
+                                     tensor_view_t<2> mean_var_tv)
+{
+    /*
+     * Each group works on a single channel.
+     * Example)
+     * x dim = {N, C, L}
+     * => gws = {C * LOCAL_SIZE}, lws = {LOCAL_SIZE}
+     *    outer_size = N, target_size = C, inner_size = L
+     */
+
+    /*
+     * Reduction to calculate mean and variance
+     */
+    const uint64_t gid = blockIdx.x;
+    const uint64_t lid = threadIdx.x;
+
+    __shared__ FLOAT_ACCUM ltmp1[LOCAL_SIZE];
```

Comment:
Originally ltmp1 and ltmp2 were single vector2 array in modnn. It might casue a little performance loss because reading and writing the vector is better in most cases.
But let's just keep it as is and fix later when we can because it's still better than current rocm.